helix-term/src/commands.rs

```diff
@@ -5595,3 +5601,153 @@ fn replay_macro(cx: &mut Context) {
+            return;
+        }
+        cx.on_next_key(move |cx, event| {
+            doc_mut!(cx.editor, &doc).remove_jump_labels(view);
+            let alphabet = &cx.editor.config().jump_label_alphabet;
+            let Some(inner ) = event.char().and_then(|ch| alphabet.iter().position(|&it| it == ch)) else {
+                return;
+            };
+            if let Some(mut range) = labels.get(outer + inner).copied() {
+                range = if behaviour == Movement::Extend {
+                    let anchor = if range.anchor < range.head {
+                        let from = primary_selection.from();
+                        if range.anchor < from {
+                            range.anchor
+                        } else {
+                            from
+                        }
+                    } else {
+                        let to = primary_selection.to();
+                        if range.anchor > to {
+                            range.anchor
+                        } else {
+                            to
+                        }
+                    };
+                     Range::new(anchor, range.head)
+                }else{
+                    range.with_direction(Direction::Forward)
+                };
+                doc_mut!(cx.editor, &doc).set_selection(view, range.into());
+            }
+        });
+    });
+}
+
+fn jump_to_word(cx: &mut Context, behaviour: Movement) {
+    // Calculate the jump candidates: ranges for any visible words with two or
+    // more characters.
+    let alphabet = &cx.editor.config().jump_label_alphabet;
+    let jump_label_limit = alphabet.len() * alphabet.len();
+    let mut words = Vec::with_capacity(jump_label_limit);
+    let (view, doc) = current_ref!(cx.editor);
+    let text = doc.text().slice(..);
+
+    // This is not necessarily exact if there is virtual text like soft wrap.
+    // It's ok though because the extra jump labels will not be rendered.
+    let start = text.line_to_char(text.char_to_line(view.offset.anchor));
+    let end = text.line_to_char(view.estimate_last_doc_line(doc) + 1);
+
+    let primary_selection = doc.selection(view.id).primary();
+    let cursor = primary_selection.cursor(text);
+    let cursor_fwd = movement::move_prev_word_start(text, Range::point(cursor), 1);
+    let mut cursor_fwd = if cursor_fwd.anchor == cursor {
+        Range::point(cursor_fwd.head)
+    } else {
+        Range::point(cursor)
+    };
+    let mut cursor_rev = cursor_fwd;
+    loop {
+        let mut changed = false;
+        if cursor_fwd.head < end {
+            cursor_fwd = movement::move_next_word_end(text, cursor_fwd, 1);
+            let range = movement::move_prev_word_start(text, cursor_fwd, 1);
+            // The cursor is on a word longer than 2 characters.
+            if RopeGraphemes::new(text.slice(range.head..))
+                .take(2)
+                .all(|g| g.chars().all(char_is_word))
```

Comment:
Ah looks like we have some duplicates in the ranges:

<details><summary>Ranges from the reproduction case...</summary>

```rust
[
	Range { anchor: 19, head: 25, .. },
	Range { anchor: 15, head: 9, .. },
	Range { anchor: 19, head: 25, .. },
	Range { anchor: 27, head: 31, .. },
	Range { anchor: 27, head: 31, .. },
	Range { anchor: 5, head: 0, .. },
	Range { anchor: 32, head: 38, .. },
	Range { anchor: 32, head: 39, .. },
	Range { anchor: 42, head: 50, .. },
	Range { anchor: 42, head: 50, .. },
	Range { anchor: 52, head: 57, .. },
	Range { anchor: 52, head: 58, .. },
	Range { anchor: 60, head: 65, .. },
	Range { anchor: 66, head: 76, .. },
	Range { anchor: 66, head: 77, .. },
	Range { anchor: 80, head: 86, .. },
	Range { anchor: 80, head: 86, .. },
	Range { anchor: 88, head: 92, .. },
	Range { anchor: 88, head: 92, .. },
	Range { anchor: 93, head: 98, .. },
	Range { anchor: 93, head: 99, .. },
	Range { anchor: 101, head: 105, .. },
	Range { anchor: 106, head: 110, .. },
	Range { anchor: 106, head: 110, .. },
	Range { anchor: 111, head: 117, .. },
	Range { anchor: 111, head: 118, .. },
	Range { anchor: 120, head: 124, .. },
	Range { anchor: 125, head: 129, .. },
	Range { anchor: 125, head: 129, .. },
	Range { anchor: 130, head: 134, .. },
	Range { anchor: 130, head: 135, .. },
	Range { anchor: 137, head: 141, .. },
	Range { anchor: 137, head: 141, .. }
]
```

</details>

`gwaw` there is going to a valid label, it's just not visible since that one is overlapped by another label.

It looks like the `eb`-like movement trips over punctuation and can re-select the last word

Reply:
It took some time to dig trough this but I fixed this and cleaned up some edgecases